src/proposal.mligo

```diff
@@ -88,6 +105,43 @@ let update_delegates (params, store : update_delegate_params * storage): return
   , { store with delegates = List.fold update_delegate params store.delegates }
   )
 
+
+// Unstake voter's tokens on a proposal that has already been flushed or dropped.
+// Fail if the voter did not vote on that proposal, or voter has already unfreezed, or
+// the proposal was not yet flushed nor dropped.
+let unstake_vote_one (config: config) (store , proposal_key : storage * proposal_key): storage =
+
+  // Ensure proposal is already flushed or dropped.
+  let p = fetch_proposal (proposal_key, store) in
+  let _ = if Set.mem (p.start_level, proposal_key) store.proposal_key_list_sort_by_level
+            then (failwith unstake_invalid_proposal : unit)
+            else unit in
+
+  // Check if voter exist.
+  let vote_status = match Big_map.find_opt (Tezos.sender, proposal_key) store.voters with
+      | Some v -> v
+      | None -> (failwith voter_does_not_exist : vote_status) in
+
+  // Ensure that the voter has not called `unstake_vote` before.
+  let _ =  if vote_status.is_unstaked = true
+      then (failwith voter_unstaked : unit)
+      else unit in
+
+  // Do the unstake
+  let tokens = vote_status.upvotes + vote_status.downvotes in
+  let store = unstake_tk(tokens, 0n, Tezos.sender, config.period, store) in
+
+  // Mark `is_unstaked` on voter's vote.
+  { store with
+      voters = Big_map.add (Tezos.sender, proposal_key) ({vote_status with is_unstaked = true}) store.voters
```

Comment:
I wonder if there is still a reason to maintain the voter in voters list after the voter has unstaked their token, @pasqu4le ?

Reply:
good point, I suppose that we could remove the entries here (and the `is_unstaked` field entirely), but I [asked here](https://serokell.slack.com/archives/C02KY15P5GE/p1642015165003300?thread_ts=1637159479.009800&cid=C02KY15P5GE) to be sure.

@rinn7e if that's the case I guess my [previous comment](https://github.com/tezos-commons/baseDAO/pull/313#discussion_r783180286) would be different and I would instead consider calling this `storage` field `staked_votes` instead

Reply:
Seems like the response is fine with not keeping track of the voters. I've removed `is_unstaked` and rename to `staked_votes`. This should be good now.

Since we no longer know if the voter has called the entrypoint previously, I removed the `voter_unstaked` and mention its description as part of `voter_does_not exist` instead.